services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java

```diff
@@ -397,8 +397,8 @@ private Response restartAuthenticationSessionFromCookie(RootAuthenticationSessio
             return Response.status(Response.Status.FOUND).location(redirectUri).build();
         } else {
             // Finally need to show error as all the fallbacks failed
-            event.error(Errors.INVALID_CODE);
-            return ErrorPage.error(session, authSession, Response.Status.BAD_REQUEST, Messages.INVALID_CODE);
+            event.error(Errors.DISABLED_COOKIES);
```

Comment:
The else branch handles the case where `authSession == null)`
[RestartLoginCookie.restartSession(..)](https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/protocol/RestartLoginCookie.java#L137) contains multiple code paths that could return `null` besides those cause by missing cookies.

I think you should check here whether the cookies are really missing.

Reply:
Hey @thomasdarimont,
i added a new method in the RestartLoginCookie class which tries to extract the cookie. The new error message is only used when in this method the extraction fails.

Reply:
I'd rather not change the method RestartLoginCookie.restartSession(..) since it is public API that might be used by clients and this change might brake their extensions.
```
public static AuthenticationSessionModel restartSession(KeycloakSession session, RealmModel realm,	    public static Cookie getRestartCookie(KeycloakSession session){
                                                            RootAuthenticationSessionModel rootSession, String expectedClientId) throws Exception {
```

Instead of changing the restartSession(..) Method you could just call your new method `getRestartCookie` within `restartSession(..)`. This would do what you need without breaking others code :)

Reply:
Hey @thomasdarimont,
are you sure that this is how it would work at this point? The restartSession() method returns either null or the AuthenticatedSessionModel. In the calling method I cannot distinguish why the restartSession() method had an error and returned null. The topic with the public API makes total sense, I had not thought about that before. I might have to change the error handling in the restartSession() method. According to the declaration it should throw an exception, but there is no exception built in. Would that be a better way? I think the restartAuthenticationSessionFromCookie() method also logs a ServiceLogger exception in the try/catch block but I don't think this is reached because the restartSession method always returns null or the AuthenticatedSessionModel.

greetings David